File: Graph/java/DUG_AdjList.java
class DU_Edge {
    int vertex;
    DU_Edge next;

    DU_Edge(int dest) {
        vertex = dest;
        next = null;
    }
}

public class DUG_AdjList {

    // Add a directed edge: src -> dest
    public static void addEdgeDUG(DU_Edge[] adjList, int src, int dest) {
        DU_Edge head = adjList[src];
        DU_Edge newEdge = new DU_Edge(dest);
        newEdge.next = head;
        adjList[src] = newEdge;
    }

    public static void printAdjList(DU_Edge[] adjList) {
        for (int i = 0; i < adjList.length; i++) {
            System.out.print(i + " : ");
            DU_Edge temp = adjList[i];
            while (temp != null) {
                System.out.print(temp.vertex);
                if (temp.next != null) System.out.print(" , ");
                temp = temp.next;
            }
            System.out.println();
        }
    }

    // Out-degree: number of neighbors in list i
    public static int countOutDegree(DU_Edge[] adjList, int target) {
        int count = 0;
        DU_Edge temp = adjList[target];
        while (temp != null) {
            count++;
            temp = temp.next;
        }
        return count;
    }

    // In-degree: count how many lists contain 'vertex'
    public static int countInDegree(DU_Edge[] adjList, int target) {
        int count = 0;
        for (int i = 0; i < adjList.length; i++) {
            DU_Edge temp = adjList[i];
            while (temp != null) {
                if (temp.vertex == target) count++;
                temp = temp.next;
            }
        }
        return count;
    }

    public static void main(String[] args) {
        int n = 6;
        DU_Edge[] adjList = new DU_Edge[n];

        addEdgeDUG(adjList, 0, 1);
        addEdgeDUG(adjList, 0, 4);
        addEdgeDUG(adjList, 1, 5);
        addEdgeDUG(adjList, 2, 3);
        addEdgeDUG(adjList, 2, 4);
        addEdgeDUG(adjList, 4, 5);
        addEdgeDUG(adjList, 1, 2);
        addEdgeDUG(adjList, 4, 1);

        printAdjList(adjList);

        int v = 2;
        System.out.println("In-degree(" + v + ")  = " + countInDegree(adjList, v));
        System.out.println("Out-degree(" + v + ") = " + countOutDegree(adjList, v));
    }
}

File: Graph/python/DUG_AdjList.py
class DU_Edge:
    def __init__(self, dest):
        self.vertex = dest
        self.next = None


def addEdgeDUG(adjList, src, dest):
    head = adjList[src]
    newEdge = DU_Edge(dest)
    newEdge.next = head
    adjList[src] = newEdge


def printAdjList(adjList):
    for i in range(len(adjList)):
        temp = adjList[i]
        print(i,':',end=' ')
        while temp is not None:
            if temp.next is not None:
                print(temp.vertex, end=', ')
            else:
                print(temp.vertex, end='')
            temp = temp.next
        print()


def countOutDegree(adjList, target):
    count = 0
    temp = adjList[target]
    while temp is not None:
        count += 1
        temp = temp.next
    return count


def countInDegree(adjList, target):
    count = 0
    i = 0
    while i < len(adjList):
        temp = adjList[i]
        while temp is not None:
            if temp.vertex == target:
                count += 1
            temp = temp.next
        i += 1
    return count


n = 6
adjList = [None] * n

addEdgeDUG(adjList, 0, 1)
addEdgeDUG(adjList, 0, 4)
addEdgeDUG(adjList, 1, 5)
addEdgeDUG(adjList, 2, 3)
addEdgeDUG(adjList, 2, 4)
addEdgeDUG(adjList, 4, 5)
addEdgeDUG(adjList, 1, 2)
addEdgeDUG(adjList, 4, 1)

printAdjList(adjList)

v = 2
print("In-degree(" + str(v) + ")  = " + str(countInDegree(adjList, v)))
print("Out-degree(" + str(v) + ") = " + str(countOutDegree(adjList, v)))


